File: C++/LinkedList_Medium_SortList_148.cpp
//Brute-Force Approach : My Approach

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) 
    {
        vector<int> vec_1;
        ListNode* temp = head;
        ListNode* prev;
        ListNode* start;

        while(temp != NULL)
        {
            vec_1.push_back(temp -> val);
            temp = temp -> next;
        }

        sort(vec_1.begin(), vec_1.end());

        if(vec_1.size() == 0)
        {
            return NULL;
        }
        else
        {
            prev = new ListNode(vec_1[0]);
            start = prev;
        }
        for(int i = 1; i < vec_1.size(); i++)
        {
            ListNode* temp = new ListNode(vec_1[i]);
            prev -> next = temp;
            prev = temp;
        }
        
        return start;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* findMiddle(ListNode* head)
    {
        ListNode* slow = head;
        ListNode* fast = head -> next;

        while(fast != NULL && fast -> next != NULL)
        {
            slow = slow -> next;
            fast = fast -> next -> next;
        }

        return slow;
    }

    ListNode* mergeHalves(ListNode* leftNode, ListNode* rightNode)
    {
        ListNode* ptr = new ListNode();
        ListNode* temp = ptr;
        while(leftNode != NULL && rightNode != NULL)
        {
            if(leftNode -> val < rightNode -> val)
            {
                temp -> next = leftNode;
                temp = temp -> next;
                leftNode = leftNode -> next;
            }
            else
            {
                temp -> next = rightNode;
                temp = temp -> next;
                rightNode = rightNode -> next;
            }
        }
        if(leftNode != NULL)
        {
            temp -> next = leftNode;
        }
        else
        {
            temp -> next = rightNode;
        }
        
        ListNode* result = ptr -> next;
        delete ptr;
        return result;
    }

    ListNode* sortList(ListNode* head) 
    {
        if(head == NULL || head -> next == NULL)
        {
            return head;
        }

        ListNode* middle = findMiddle(head);
        ListNode* left = head;
        ListNode* right = middle -> next;

        middle -> next = NULL;

        ListNode* leftNode = sortList(left);
        ListNode* rightNode = sortList(right);

        return mergeHalves(leftNode, rightNode);
    }
};

File: Python/LinkedList_Medium_SortList_148.py
# Optimal Approach : 

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def findMiddle(self, head):
        slow = head
        fast = head.next

        while fast != None and fast.next != None:
            slow = slow.next
            fast = fast.next.next
        
        return slow

    def mergeHalves(self, leftNode, rightNode):
        ptr = ListNode()
        temp = ptr

        while leftNode != None and rightNode != None:
            if leftNode.val < rightNode.val:
                temp.next = leftNode
                temp = temp.next
                leftNode = leftNode.next
            else:
                temp.next = rightNode
                temp = temp.next
                rightNode = rightNode.next
            
        if leftNode == None:
            temp.next = rightNode
        else:
            temp.next = leftNode

        return ptr.next

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if head == None or head.next == None:
            return head

        left = head
        middle = self.findMiddle(head)
        right = middle.next
        middle.next = None

        leftNode = self.sortList(left)
        rightNode = self.sortList(right)

        return self.mergeHalves(leftNode, rightNode)
        